subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1236,31 +1244,126 @@ ssize_t ztls_recv(int sock, void *buf, size_t max_len, int flags)
 	return ztls_recvfrom(sock, buf, max_len, flags, NULL, 0);
 }
 
+static ssize_t sendto_tls(struct net_context *context, const void *buf,
+			  size_t len, int flags)
+{
+	int ret;
+
+	ret = mbedtls_ssl_write(&context->tls->ssl, buf, len);
+	if (ret >= 0) {
+		return ret;
+	}
+
+	if (ret == MBEDTLS_ERR_SSL_WANT_READ ||
+	    ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+		errno = EAGAIN;
+	} else {
+		errno = EIO;
+	}
+
+	return -1;
+}
+
+#if defined(CONFIG_NET_SOCKETS_ENABLE_DTLS)
+static ssize_t sendto_dtls_client(struct net_context *context, const void *buf,
+				  size_t len, int flags,
+				  const struct sockaddr *dest_addr,
+				  socklen_t addrlen)
+{
+	int ret;
+
+	if (!dest_addr) {
+		/* No address provided, check if we have stored one,
+		 * otherwise return error.
+		 */
+		if (context->tls->dtls_peer_addrlen == 0) {
+			ret = -EDESTADDRREQ;
+			goto error;
+		}
+	} else if (context->tls->dtls_peer_addrlen == 0) {
+		/* Address provided and no peer address stored. */
+		dtls_peer_address_set(context, dest_addr, addrlen);
+	} else if (!dtls_is_peer_addr_valid(context, dest_addr, addrlen) != 0) {
+		/* Address provided but it does not match stored one */
+		ret = -EISCONN;
+		goto error;
+	}
+
+	if (!context->tls->is_initialized) {
+		ret = tls_mbedtls_init(context, false);
+		if (ret < 0) {
+			goto error;
+		}
+	}
+
+	if (!context->tls->tls_established) {
+		ret = tls_mbedtls_handshake(context);
+		if (ret < 0) {
+			goto error;
+		}
+	}
+
+	return sendto_tls(context, buf, len, flags);
+
+error:
+	errno = -ret;
+	return -1;
+}
+
+static ssize_t sendto_dtls_server(struct net_context *context, const void *buf,
+				  size_t len, int flags,
+				  const struct sockaddr *dest_addr,
+				  socklen_t addrlen)
+{
+	/* For DTLS server, require to have established DTLS connection
+	 * in order to send data.
+	 */
+	if (!context->tls->tls_established) {
+		errno = ENOTCONN;
+		return -1;
+	}
+
+	/* Check if not trying to send to other peer, which we're
```

Comment:
This comment is hard to read, because it has double negation ;-).

Reply:
Agreed, it might not be straightforward, will repharse.